File: ylcs-app/iosApp/core/PIPView.swift
import Foundation
import UIKit
import AVKit

@objc
public protocol PictureInPictureDelegate where Self: AnyObject {
    
    /// Start PIP.
    func didStartPictureInPicture()
    
    /// Stop PIP.
    func didStopPictureInPicture()
    
    /// PIP failed to start.
    func didFailToStartPictureInPicture(error: Error)
}

@objcMembers
public class PIPView: UIView, AVPictureInPictureSampleBufferPlaybackDelegate, AVPictureInPictureControllerDelegate {

    public override class var layerClass: AnyClass {
        AVSampleBufferDisplayLayer.self
    }
    
    public weak var delegate: PictureInPictureDelegate?
    
    public var pipController: AVPictureInPictureController?
        
    public func setupPIP() {
        if (pipController != nil) {
            return
        }
        pipController = AVPictureInPictureController(contentSource: .init(sampleBufferDisplayLayer: self.layer as! AVSampleBufferDisplayLayer, playbackDelegate: self))
        pipController?.delegate = self
        pipController?.setValue(1, forKey: "controlsStyle")
    }
    
    public func getSnapshot() -> CMSampleBuffer {
        let view = self.subviews.first!
        return createSampleBufferFrom(pixelBuffer: view.asImage().asBuffer()!)!
    }
    
    public func pictureInPictureController(_ pictureInPictureController: AVPictureInPictureController, setPlaying playing: Bool) {
    }
    
    public func pictureInPictureControllerTimeRangeForPlayback(_ pictureInPictureController: AVPictureInPictureController) -> CMTimeRange {
        CMTimeRange(start: .indefinite, duration: CMTime(seconds: .infinity, preferredTimescale: CMTimeScale(NSEC_PER_SEC)))
    }
    
    public func pictureInPictureControllerIsPlaybackPaused(_ pictureInPictureController: AVPictureInPictureController) -> Bool {
        false
    }
    
    public func pictureInPictureController(_ pictureInPictureController: AVPictureInPictureController, didTransitionToRenderSize newRenderSize: CMVideoDimensions) {
    }
    
    public func pictureInPictureController(_ pictureInPictureController: AVPictureInPictureController, skipByInterval skipInterval: CMTime, completion completionHandler: @escaping () -> Void) {
    }
    
    public func pictureInPictureController(_ pictureInPictureController: AVPictureInPictureController, failedToStartPictureInPictureWithError error: Error) {
        delegate?.didFailToStartPictureInPicture(error: error)
    }
    
    public func pictureInPictureControllerDidStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        delegate?.didStartPictureInPicture()
    }
    
    public func pictureInPictureControllerDidStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        delegate?.didStopPictureInPicture()
    }
}

File: ylcs-module/compose/core/src/iosMain/kotlin/love/yinlin/compose/window/PIPView.kt
@file:OptIn(ExperimentalForeignApi::class)
package love.yinlin.compose.window

import kotlinx.cinterop.*
import platform.AVFoundation.*
import platform.AVKit.*
import platform.CoreGraphics.*
import platform.CoreMedia.*
import platform.Foundation.*
import platform.UIKit.*
import platform.darwin.NSEC_PER_SEC

interface PictureInPictureDelegate {
    fun didStartPictureInPicture()
    fun didStopPictureInPicture()
    fun didFailToStartPictureInPicture(error: NSError)
}

class PIPView(frame: CValue<CGRect>) : UIView(frame),
    AVPictureInPictureSampleBufferPlaybackDelegateProtocol,
    AVPictureInPictureControllerDelegateProtocol {
    val displayLayer: AVSampleBufferDisplayLayer get() = this.layer as AVSampleBufferDisplayLayer
    var delegate: PictureInPictureDelegate? = null
    var pipController: AVPictureInPictureController? = null

    @OptIn(BetaInteropApi::class)
    fun setupPIP() {
        if (pipController != null) return

        val contentSource = AVPictureInPictureControllerContentSource()
        AVPictureInPictureControllerContentSource.create(displayLayer, this)

        pipController = AVPictureInPictureController(contentSource = contentSource).also {
            it.delegate = this
            it.setValue(NSNumber(int = 1), forKey = "controlsStyle")
        }
    }

    fun getSnapshot(): CMSampleBufferRef? {
        val subview = subviews.firstOrNull() as? UIView ?: return null
        return subview.asImage().asSampleBuffer()
    }

    override fun pictureInPictureController(pictureInPictureController: AVPictureInPictureController, setPlaying: Boolean) {

    }

    override fun pictureInPictureControllerTimeRangeForPlayback(pictureInPictureController: AVPictureInPictureController): CValue<CMTimeRange> {
        return CMTimeRangeMake(
            start = kCMTimeIndefinite.readValue(),
            duration = CMTimeMakeWithSeconds(Double.MAX_VALUE, NSEC_PER_SEC.toInt())
        )
    }

    override fun pictureInPictureControllerIsPlaybackPaused(pictureInPictureController: AVPictureInPictureController): Boolean {
        return false
    }

    override fun pictureInPictureController(pictureInPictureController: AVPictureInPictureController, didTransitionToRenderSize: CValue<CMVideoDimensions>) {

    }

    override fun pictureInPictureController(pictureInPictureController: AVPictureInPictureController, skipByInterval: CValue<CMTime>, completionHandler: () -> Unit) {
        completionHandler()
    }

    override fun pictureInPictureController(pictureInPictureController: AVPictureInPictureController, failedToStartPictureInPictureWithError: NSError) {
        delegate?.didFailToStartPictureInPicture(failedToStartPictureInPictureWithError)
    }

    override fun pictureInPictureControllerDidStartPictureInPicture(pictureInPictureController: AVPictureInPictureController) {
        delegate?.didStartPictureInPicture()
    }

    override fun pictureInPictureControllerDidStopPictureInPicture(pictureInPictureController: AVPictureInPictureController) {
        delegate?.didStopPictureInPicture()
    }
}